### PR TITLE
refactor: add DL model class

### PR DESCRIPTION
**Pull request by ylwu-amzn**
Signed-off-by: Yaliang Wu <ylwu@amazon.com>

### Description
Refactor code to make it more extensible.
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: common/src/main/java/org/opensearch/ml/common/FunctionName.java [hunk 1]
```diff
@@ -25,4 +25,15 @@ public static FunctionName from(String value) {
             throw new IllegalArgumentException("Wrong function name");
         }
     }
+
+    /**
+     * Check if model is DL model.
+     * @return true for DL model.
+     */
+    public static boolean isDLModel(FunctionName functionName) {
```
[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply common/src/main/java/org/opensearch/ml/common/FunctionName.java:33]: This is for future extension. Not just sentence embedding models, we will add more models

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/DLModel.java [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+        try {
+            return AccessController.doPrivileged((PrivilegedExceptionAction<ModelTensorOutput>) () -> {
+                Thread.currentThread().setContextClassLoader(getClass().getClassLoader());
+                if (predictors == null) {
+                    throw new MLException("model not loaded.");
+                }
+                return predict(modelId, mlInput.getInputDataset());
+            });
+        } catch (Throwable e) {
+            String errorMsg = "Failed to inference text embedding";
```
[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/DLModel.java:81]: Good catch

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/DLModel.java [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+            AccessController.doPrivileged((PrivilegedExceptionAction<Void>) () -> {
+                ClassLoader contextClassLoader = Thread.currentThread().getContextClassLoader();
+                try {
+                    System.setProperty("PYTORCH_PRECXX11", "true");
+                    System.setProperty("DJL_CACHE_DIR", mlEngine.getDjlCachePath().toAbsolutePath().toString());
+                    // DJL will read "/usr/java/packages/lib" if don't set "java.library.path". That will throw
+                    // access denied exception
+                    System.setProperty("java.library.path", mlEngine.getDjlCachePath().toAbsolutePath().toString());
+                    System.setProperty("ai.djl.pytorch.num_interop_threads", "1");
+                    System.setProperty("ai.djl.pytorch.num_threads", "1");
```
[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/DLModel.java:181]: yes, this is from our performance testing

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/DLModel.java [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+                    }
+                    if (predictorList.size() > 0) {
+                        this.predictors = predictorList.toArray(new Predictor[0]);
+                        predictorList.clear();
+                    }
+                    if (modelList.size() > 0) {
+                        this.models = modelList.toArray(new ZooModel[0]);
+                        modelList.clear();
+                    }
+                    log.info("Load model {} successfully on {} devices", modelId, devices.length);
```
[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/DLModel.java:248]: sounds good

File: common/src/main/java/org/opensearch/ml/common/FunctionName.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (dhrubo-os) [reply common/src/main/java/org/opensearch/ml/common/FunctionName.java:33]: Can we please add a comment then for better clarification?

[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply common/src/main/java/org/opensearch/ml/common/FunctionName.java:33]: Actually I don't add to keep not confuse community user as we don't have clear plan yet like add which type of model and when. I would keep it as is.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/DLModel.java [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+                        modelList.clear();
+                    }
+                    throw new MLException(errorMessage, e);
+                } finally {
+                    deleteFileQuietly(mlEngine.getLoadModelPath(modelId));
+                    Thread.currentThread().setContextClassLoader(contextClassLoader);
+                }
+            });
+        } catch (PrivilegedActionException e) {
+            String errorMsg = "Failed to load model";
```
Comment (dhrubo-os) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/DLModel.java:269]: [minor] Can we please print the model id here also when you raise another PR?

**Review by dhrubo-os — DISMISSED** [review]

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/DLModel.java (hunk 5 above)
Comment (ylwu-amzn) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/DLModel.java]: Fixed together with some failed UT

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/DLModel.java [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+                engine = PYTORCH_ENGINE;
+                break;
+            case ONNX:
+                engine = ONNX_ENGINE;
+                break;
+            default:
+                throw new IllegalArgumentException("unsupported engine");
+        }
+
+        File modelZipFile = (File)params.get(MODEL_ZIP_FILE);
```
Comment (Zhangxunmt) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/DLModel.java:112]: This model only support initialization from a local zip file in the ml-common host? I think it should also download file from url and load?

**Review by rbhavna — APPROVED** [review]

**Review by Zhangxunmt — APPROVED** [review]

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/DLModel.java (hunk 6 above)
Comment (ylwu-amzn) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/DLModel.java:112]: Download from URL is a step before this part. That means model already downloaded to local before this line.

**Pull request merged**